bfvmm/include/hve/arch/intel_x64/ept/memory_map.h

```diff
@@ -0,0 +1,172 @@
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+
+#ifndef MEMORY_MAP_EPT_INTEL_X64_H
+#define MEMORY_MAP_EPT_INTEL_X64_H
+
+#include <bfgsl.h>
+#include <bftypes.h>
+#include <bfmemory.h>
+#include <mutex>
+
+#include "intrinsics.h"
+#include "types.h"
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_EAPIS_HVE
+#ifdef SHARED_EAPIS_HVE
+#define EXPORT_EAPIS_HVE EXPORT_SYM
+#else
+#define EXPORT_EAPIS_HVE IMPORT_SYM
+#endif
+#else
+#define EXPORT_EAPIS_HVE
+#endif
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+// *INDENT-OFF*
+
+namespace eapis
+{
+namespace intel_x64
+{
+namespace ept
+{
+
+class EXPORT_EAPIS_HVE memory_map
+{
+public:
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    memory_map();
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    ~memory_map();
+
+    /// Map guest physical address to host physical address with the given
+    /// page size
+    ///
+    /// @expects hpa aligned to size
+    /// @ensures
+    ///
+    /// @return Returns the leaf extended page table entry that maps gpa->hpa
+    ///
+    epte_t & map(gpa_t gpa, hpa_t hpa, uint64_t size) const;
+
+    /// Unmap a page by guest physical address
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    void unmap(gpa_t gpa) const;
+
+    /// Guest physical address to leaf extended page table entry
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the leaf extended page table entry that maps gpa->hpa
+    ///
+    epte_t & gpa_to_epte(gpa_t gpa) const;
+
+    /// Guest physical address to host physical address
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the host physical address that gpa is mapped to
+    ///
+    hpa_t gpa_to_hpa(gpa_t gpa) const;
+
+    /// Convert this memory maps page tables to a flat memory descriptor list.
+    /// NOTE: The returned memory descriptor list does not describe memory
+    /// mapped by the page tables, but rather the memory used to hold the
+    /// extended page tables themselves.
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return List of memory descriptors that describe the extended page
+    /// tables
+    ///
+    std::vector<memory_descriptor> to_mdl() const;
+
+    /// Return the base host physical address of this memory map
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return This memory maps base host physical address
+    ///
+    hpa_t hpa() const;
+
+#ifndef ENABLE_BUILD_TEST
+private:
+#endif
+    hva_t m_pml4_hva{0};
+    hpa_t m_pml4_hpa{0};
+    mutable std::mutex m_mutex;
```

Comment:
To be more specific, mods to EPT require TLB flushing, which is not handled by Intel automatically in a multicore environment, so if you have a global EPT that you plan to modify, you would have to send IPIs to the other cores to flush their TLBs prior to any potential accesses, which is hard to get right. Basically, you have to "pause" the VM on all cores and wait for an IPI, make the EPT mod, send your IPIs to the other cores, flush, and resume. Real pain. In most cases, EPT is setup and mods to EPT are made on a per core basis to prevent this issue. Only in the case of shared memory between guest VMs do we have to go down this rabbit hole.

Reply:
There was a mutex in the previous code. My thoughts while developing this were that a mutex is needed because the memory map relies on all page tables remaining constant while a page table walk is happening. If a global memory map were being manipulated by multiple threads, you could run into problems if one thread was trying to resolve a mapping while another thread tries to delete one (even if they weren't the same gpa).

With that said, you bring up a good point. Just slapping a mutex on these functions to serialize access to the memory map does not even come close to solving all issues of synchronization between cores. I could see the case for removing the mutex from the memory map, and instead letting some other higher level component manage all the complexities of synchronization.

I'd be curious if @connojd and @tklengyel have any thoughts about this.

Reply:
If a mutex was there, it should have been removed. 

Once again, a mutex provides no value as the other cores have to be in a wait-for-IPI state to make any mods to EPT if you are using a global EPT, which is why EPTP switching was added to the Intel architecture. 

Oddly, this is not an issue on ARM because you can globally flush the TLBs on all cores from a single instruction, which you cannot do on Intel.